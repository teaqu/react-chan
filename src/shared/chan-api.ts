import { Thread } from 'src/thread/thread';
import { Post } from 'src/post/post';

export interface ChanAPI {
  fetchCatalog(boardId: string): Promise<Thread[]>;
  fetchThread(boardId: string, threadNo: number): Promise<Post[]>;
}
