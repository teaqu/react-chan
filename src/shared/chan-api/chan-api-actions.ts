import { createAction } from '@reduxjs/toolkit';

const switchChanAPI = createAction<string>('SWITCH_CHAN_API');

export default {
  switchChanAPI
};
