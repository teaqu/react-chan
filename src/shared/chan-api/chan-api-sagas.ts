import { AnyAction } from 'redux';
import { setContext, takeLatest } from 'redux-saga/effects';

import actions from './chan-api-actions';
import fourChanAPI from './four-chan-api';

//
export function* switchChanAPI(action: AnyAction) {
  if (action.payload === '4chan') {
    setContext(fourChanAPI);
  }
}

export function* watchFetchCatalog() {
  yield takeLatest(actions.switchChanAPI, switchChanAPI);
}
