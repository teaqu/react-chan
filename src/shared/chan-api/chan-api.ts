import { Thread } from 'src/thread/thread';
import { Post } from 'src/post/post';
import { Board } from 'src/board/board';

// So we can support more image boards in the future...
export interface ChanAPI {
  fetchCatalog(boardId: string): Promise<Threads>;
  fetchThread(boardId: string, threadNo: number): Promise<Posts>;
  fetchBoards(): Promise<Board[]>;
  calcReplies(posts: Posts): Posts;
}

export interface Posts {
  [no: string]: Post;
}

export interface Threads {
  [no: string]: Thread;
}
