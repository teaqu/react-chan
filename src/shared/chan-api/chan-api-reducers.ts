import { createReducer } from '@reduxjs/toolkit';

import actions from './chan-api-actions';
import fourChanAPI from './four-chan-api';
import { ChanAPI } from './chan-api';

const initialState: ChanAPI = fourChanAPI;

export default createReducer(initialState, {
  [actions.switchChanAPI.type]: (state, action) => {
    if (action.payload === '4chan') {
      // eslint-disable-next-line no-param-reassign
      state = fourChanAPI;
    }
  }
});
