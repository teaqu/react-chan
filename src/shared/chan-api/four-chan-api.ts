import { Thread } from 'src/thread/thread';
import { Post } from 'src/post/post';
import { Board } from 'src/board/board';

import { ChanAPIState } from './chan-api-reducers';
import { Posts, Threads } from './chan-api';

interface CatalogAPI {
  page: number;
  threads: Thread[];
}

interface ThreadAPI {
  posts: Post[];
}

interface BoardAPI {
  boards: Board[];
  troll_flags: {
    [flag: string]: string;
  };
}

export const FourChanState: ChanAPIState = {
  thumbnail: 'https://i.4cdn.org/[board]/[tim]s.jpg',
  image: 'https://i.4cdn.org/[board]/[tim]',
  fileDeleted: 'https://s.4cdn.org/image/filedeleted.gif',
  flag: 'https://s.4cdn.org/image/country/[country].gif',
  trollFlag: 'https://s.4cdn.org/image/country/troll/[country].gif',
  since4pass: 'https://s.4cdn.org/image/minileaf.gif'
};

export const fetchCatalog = (boardId: string): Promise<Threads> => {
  const uri = `https://a.4cdn.org/${boardId}/catalog.json`;
  return fetch(uri)
    .then(response => response.json())
    .catch(error => error)
    .then((catalogs: CatalogAPI[]) =>
      // Normalise the catalogs into threads with their current page.
      catalogs.flatMap((catalog: CatalogAPI) =>
        catalog.threads.map((thread: Thread) => ({
          ...thread,
          page: catalog.page
        }))
      )
    )
    .then(threads =>
      threads.reduce(
        (acc, curr) => ({ ...acc, [curr.no.toString()]: curr }),
        {}
      )
    );
};

export const fetchThread = (
  boardId: string,
  threadNo: number
): Promise<Posts> => {
  const uri = `https://a.4cdn.org/${boardId}/thread/${threadNo}.json`;
  return fetch(uri)
    .then(response => response.json())
    .catch(error => error)
    .then((thread: ThreadAPI) => thread.posts)
    .then(posts =>
      // Set posts array to no => post key value object
      posts.reduce((acc, curr) => ({ ...acc, [curr.no]: { ...curr } }), {})
    );
};

export const fetchBoards = (): Promise<Board[]> => {
  const uri = 'https://a.4cdn.org/boards.json';
  return fetch(uri)
    .then(response => response.json())
    .catch(error => error)
    .then((api: BoardAPI) => api.boards);
};

/**
 * Calculate replies for our post
 * As this must be calculated before we show can show the thread,
 * it must use as few loops and regex as possible.
 * @param posts
 */
export const calcReplies = (posts: Posts): Posts => {
  const regex = /class="quotelink">&gt;&gt;(.*?)</g; // look for >>(postNo)
  let matches = null;
  // Find all replies within each post
  for (let no in posts) {
    posts[no].reply_links = [];
    do {
      matches = regex.exec(posts[no].com);
      if (matches) {
        const match = parseInt(matches[1], 10);

        // Add this post to the reply links of the post it's replying to.
        if (posts[match]) {
          posts[match].reply_links.push(posts[no].no);
        }
      }
    } while (matches);
  }
  return posts;
};

export default { fetchCatalog, fetchThread, fetchBoards, calcReplies };
