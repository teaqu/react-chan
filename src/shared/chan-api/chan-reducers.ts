import { createReducer } from '@reduxjs/toolkit';

import actions from './chan-actions';
import { FourChanState } from './four-chan-api';

export interface ChanAPIState {
  thumbnail: string;
  image: string;
  fileDeleted: string;
}

const initialState = FourChanState;

export default createReducer(initialState, {
  [actions.switchChanAPI.type]: (state, action) => {
    if (action.payload === '4chan') {
      state = FourChanState;
    }
  }
});
