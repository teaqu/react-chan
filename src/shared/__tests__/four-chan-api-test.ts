import fetchMock from 'fetch-mock';

import FourChanAPI from '../chan-api/four-chan-api';
import { Threads } from '../chan-api/chan-api';

import catalogResponse from './__mocks__/catalog-response.json';

describe('catalog reducer', () => {
  afterEach(() => {
    fetchMock.restore();
  });
  it('should return the initial state', async () => {
    const boardId = 'a';
    fetchMock.getOnce(`https://a.4cdn.org/${boardId}/catalog.json`, {
      body: catalogResponse,
      headers: { 'content-type': 'application/json' }
    });
    const threads: Threads = await FourChanAPI.fetchCatalog(boardId);
    expect(threads).toMatchSnapshot();
  });
});
