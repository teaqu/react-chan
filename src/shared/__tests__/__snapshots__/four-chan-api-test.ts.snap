// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`catalog reducer should return the initial state 1`] = `
Object {
  "200705264": Object {
    "bumplimit": 0,
    "com": "I miss this pig.",
    "custom_spoiler": 1,
    "ext": ".jpg",
    "filename": "images (2)",
    "fsize": 38395,
    "h": 586,
    "imagelimit": 0,
    "images": 20,
    "last_modified": 1585155132,
    "last_replies": Array [
      Object {
        "com": "<a href=\\"#p200719748\\" class=\\"quotelink\\">&gt;&gt;200719748</a><br>non-ambiguous ends are a death sentence for harem novels and manga<br>assmad nips wont buy the book out of protest that their best girl didnt win, and the rest wont buy the book because they cant relate to a non-virgin mc anymore",
        "name": "Anonymous",
        "no": 200719833,
        "now": "03/25/20(Wed)12:11:50",
        "resto": 200705264,
        "time": 1585152710,
      },
      Object {
        "com": "krulchads rise up",
        "ext": ".webm",
        "filename": "krul takes what she wants longer",
        "fsize": 2772590,
        "h": 720,
        "md5": "zzqW6zfJsBQJze/n1nDh7A==",
        "name": "Anonymous",
        "no": 200719962,
        "now": "03/25/20(Wed)12:15:17",
        "resto": 200705264,
        "tim": 1585152917145,
        "time": 1585152917,
        "tn_h": 70,
        "tn_w": 125,
        "w": 1280,
      },
      Object {
        "com": "<a href=\\"#p200719833\\" class=\\"quotelink\\">&gt;&gt;200719833</a><br>I don&#039;t blame them. Harem LNs and manga should either end ambiguously or with a harem.",
        "name": "Anonymous",
        "no": 200719970,
        "now": "03/25/20(Wed)12:15:34",
        "resto": 200705264,
        "time": 1585152934,
      },
      Object {
        "com": "<a href=\\"#p200716406\\" class=\\"quotelink\\">&gt;&gt;200716406</a><br>Bahamut, Antimagic, Blade Dance, Fafnir best uniforms.<br><s>Fafnir last 2 volumes translation when? Hell, the other 2 are already finished, with Bahamut following after. I just want to know if Yuu will get his harem in the end.</s>",
        "ext": ".jpg",
        "filename": "1450585916229",
        "fsize": 948502,
        "h": 1394,
        "md5": "mKH/fwG4lsi0+cC6+8tW8A==",
        "name": "Anonymous",
        "no": 200720055,
        "now": "03/25/20(Wed)12:18:09",
        "resto": 200705264,
        "tim": 1585153089900,
        "time": 1585153089,
        "tn_h": 125,
        "tn_w": 113,
        "w": 1271,
      },
      Object {
        "com": "<a href=\\"#p200719962\\" class=\\"quotelink\\">&gt;&gt;200719962</a><br>Good times<br>I miss Bahamut threads",
        "ext": ".jpg",
        "filename": "1458581436825",
        "fsize": 178182,
        "h": 1083,
        "md5": "FAJJshrKsF2I2JEGqvyMkw==",
        "name": "Anonymous",
        "no": 200721251,
        "now": "03/25/20(Wed)12:52:12",
        "resto": 200705264,
        "tim": 1585155132161,
        "time": 1585155132,
        "tn_h": 125,
        "tn_w": 88,
        "w": 766,
      },
    ],
    "md5": "RsUk0Tixfd77CgXVRlwwKA==",
    "name": "Anonymous",
    "no": 200705264,
    "now": "03/25/20(Wed)00:49:02",
    "omitted_images": 20,
    "omitted_posts": 66,
    "page": 2,
    "replies": 71,
    "resto": 0,
    "semantic_url": "rokudenashi",
    "sub": "Rokudenashi",
    "tim": 1585111742012,
    "time": 1585111742,
    "tn_h": 250,
    "tn_w": 223,
    "w": 523,
  },
  "200711262": Object {
    "bumplimit": 0,
    "com": "Fake Yuzaki replaces Cosmo. How would he fare in the Annihilation Tournament?",
    "custom_spoiler": 1,
    "ext": ".png",
    "filename": "T20",
    "fsize": 1034875,
    "h": 1361,
    "imagelimit": 0,
    "images": 117,
    "last_modified": 1585155132,
    "last_replies": Array [
      Object {
        "com": "<a href=\\"#p200721080\\" class=\\"quotelink\\">&gt;&gt;200721080</a><br><span class=\\"quote\\">&gt;what art did koga use to clog his veins and not bleed to death after he pulled the knives</span>",
        "name": "Anonymous",
        "no": 200721187,
        "now": "03/25/20(Wed)12:50:08",
        "resto": 200711262,
        "time": 1585155008,
      },
      Object {
        "com": "I certainly hope Koga&#039;s PTSD doesn&#039;t flare up in the next chapter.",
        "ext": ".png",
        "filename": "1571286980669",
        "fsize": 1064928,
        "h": 1361,
        "md5": "wHQr4p7ytFTdwtCwMKBbMg==",
        "name": "Anonymous",
        "no": 200721231,
        "now": "03/25/20(Wed)12:51:34",
        "resto": 200711262,
        "tim": 1585155094553,
        "time": 1585155094,
        "tn_h": 125,
        "tn_w": 88,
        "w": 960,
      },
      Object {
        "com": "<a href=\\"#p200717727\\" class=\\"quotelink\\">&gt;&gt;200717727</a><br>I can totally see it",
        "name": "Anonymous",
        "no": 200721240,
        "now": "03/25/20(Wed)12:51:49",
        "resto": 200711262,
        "time": 1585155109,
      },
      Object {
        "com": "<a href=\\"#p200721180\\" class=\\"quotelink\\">&gt;&gt;200721180</a><br><span class=\\"quote\\">&gt;so he want to go Julius way</span><br><span class=\\"quote\\">&gt;Toya electric is that way Koga</span>",
        "name": "Anonymous",
        "no": 200721243,
        "now": "03/25/20(Wed)12:51:55",
        "resto": 200711262,
        "time": 1585155115,
      },
      Object {
        "com": "<a href=\\"#p200721231\\" class=\\"quotelink\\">&gt;&gt;200721231</a><br>Koga is going off the deep end after this.",
        "name": "Anonymous",
        "no": 200721250,
        "now": "03/25/20(Wed)12:52:12",
        "resto": 200711262,
        "time": 1585155132,
      },
    ],
    "md5": "EMJDrxxnlfJvUxxTepL+mw==",
    "name": "Anonymous",
    "no": 200711262,
    "now": "03/25/20(Wed)05:54:27",
    "omitted_images": 117,
    "omitted_posts": 401,
    "page": 1,
    "replies": 406,
    "resto": 0,
    "semantic_url": "kengan-omega-asura",
    "sub": "Kengan Omega / Asura",
    "tim": 1585130067777,
    "time": 1585130067,
    "tn_h": 250,
    "tn_w": 176,
    "w": 960,
  },
  "200715263": Object {
    "bumplimit": 0,
    "com": "<span class=\\"quote\\">&gt;New Higurashi no Naku koro ni anime, with the original voice actors and more honest adaptation of the VN sprites</span><br><span class=\\"quote\\">&gt;New Higurashi no Naku koro ni manga</span><br>Too good to be true. Is this heaven?",
    "custom_spoiler": 1,
    "ext": ".jpg",
    "filename": "ETnLx67U8AArzpH",
    "fsize": 4058291,
    "h": 2866,
    "imagelimit": 0,
    "images": 21,
    "last_modified": 1585155135,
    "last_replies": Array [
      Object {
        "com": "<a href=\\"#p200720899\\" class=\\"quotelink\\">&gt;&gt;200720899</a><br>Has he ever patted Mion&#039;s head? I think that&#039;s something he only does to Shion, so there may be a part of Keiichi who can instinctively tell them apart without him actually noticing it, like in Meakashi&#039;s second ending.",
        "name": "Anonymous",
        "no": 200721014,
        "now": "03/25/20(Wed)12:45:05",
        "resto": 200715263,
        "time": 1585154705,
      },
      Object {
        "com": "<a href=\\"#p200720893\\" class=\\"quotelink\\">&gt;&gt;200720893</a><br><span class=\\"quote\\">&gt;Live2d</span><br>Weirdchamp",
        "name": "Anonymous",
        "no": 200721077,
        "now": "03/25/20(Wed)12:47:02",
        "resto": 200715263,
        "time": 1585154822,
      },
      Object {
        "com": "<a href=\\"#p200715446\\" class=\\"quotelink\\">&gt;&gt;200715446</a><br>Why is her hat flying off her head?",
        "name": "Anonymous",
        "no": 200721121,
        "now": "03/25/20(Wed)12:48:14",
        "resto": 200715263,
        "time": 1585154894,
      },
      Object {
        "com": "<a href=\\"#p200720799\\" class=\\"quotelink\\">&gt;&gt;200720799</a><br>Yeah. I&#039;m a big fan of DEENgurashi but that scene is up on Youtube as an example of &quot;shitty Japanese dub voice acting&quot; for dubfags to harp on just because the visuals are sloppily done.",
        "name": "Anonymous",
        "no": 200721245,
        "now": "03/25/20(Wed)12:52:01",
        "resto": 200715263,
        "time": 1585155121,
      },
      Object {
        "com": "<a href=\\"#p200720613\\" class=\\"quotelink\\">&gt;&gt;200720613</a><br><br>Hmm! i didn&#039;t think of it that way, yes Tsuisou no Despair does fit the first few chapters really well.<br><br>I haven&#039;t heard the Sui OP much, but it&#039;s actually real good too. It might fit the later fighting fate chapters real well too.<br><br>Gosh I love Itou Kanako&#039;s songs.<br><br>Also yes the trippy flowers, and the DEEN OPs in general are sooo good. OP1 is still one of my favorite OPs ever. The visual language is just so good. The flowers for example, are seen falling in perfect unison, which gives a very unnatural feeling. But it&#039;s symbolic of not only the way Oyashiro&#039;s curse is affecting everyone in the village, but the uniformity of it all kind of implies that everything is being planned by *someone* or something. Like an organized conspiracy feeling. Of course it also foreshadows that there is a lot of death coming. The flowers fit very well with the the setting, and the trippiness of them and how in your face they are at some points, makes you feel a bit suffocated too.<br><br>I also love the shot zooming out from the candle while its flame goes out. Gosh damn.",
        "name": "Anonymous",
        "no": 200721252,
        "now": "03/25/20(Wed)12:52:15",
        "resto": 200715263,
        "time": 1585155135,
      },
    ],
    "md5": "PZ3ERE0XBzJQmkjt3x7gVg==",
    "name": "Anonymous",
    "no": 200715263,
    "now": "03/25/20(Wed)09:39:25",
    "omitted_images": 18,
    "omitted_posts": 83,
    "page": 1,
    "replies": 88,
    "resto": 0,
    "semantic_url": "higurashi-no-naku-koro-ni",
    "sub": "&quot;Higurashi no Naku Koro ni &quot;",
    "tim": 1585143565420,
    "time": 1585143565,
    "tn_h": 250,
    "tn_w": 176,
    "w": 2024,
  },
  "200719599": Object {
    "bumplimit": 0,
    "com": "Spoilers tonight/tomorrow.<br>Are we gonna see a roasted chicken or we will cut away?<br>Also, which student would you like to get a spotlight in this arc? Kaminari and Tokoyami already got theirs.",
    "custom_spoiler": 1,
    "ext": ".png",
    "filename": "stepping on the food",
    "fsize": 382898,
    "h": 838,
    "imagelimit": 0,
    "images": 2,
    "last_modified": 1585155129,
    "last_replies": Array [
      Object {
        "com": "<a href=\\"#p200719871\\" class=\\"quotelink\\">&gt;&gt;200719871</a><br>Giga wouldn&#039;t let you ride him unless Shiggy/AfO says so",
        "name": "Anonymous",
        "no": 200720032,
        "now": "03/25/20(Wed)12:17:33",
        "resto": 200719599,
        "time": 1585153053,
      },
      Object {
        "com": "<a href=\\"#p200719782\\" class=\\"quotelink\\">&gt;&gt;200719782</a><br>How low can we go?",
        "name": "Anonymous",
        "no": 200720137,
        "now": "03/25/20(Wed)12:20:25",
        "resto": 200719599,
        "time": 1585153225,
      },
      Object {
        "com": "<a href=\\"#p200719782\\" class=\\"quotelink\\">&gt;&gt;200719782</a><br><span class=\\"quote\\">&gt;99999) Hunter x Hunter - 0</span>",
        "name": "Anonymous",
        "no": 200720490,
        "now": "03/25/20(Wed)12:30:06",
        "resto": 200719599,
        "time": 1585153806,
      },
      Object {
        "com": "<a href=\\"#p200719782\\" class=\\"quotelink\\">&gt;&gt;200719782</a><br>isn&#039;t neverland on its way out",
        "name": "Anonymous",
        "no": 200720965,
        "now": "03/25/20(Wed)12:43:35",
        "resto": 200719599,
        "time": 1585154615,
      },
      Object {
        "com": "Oh, so those spoilers from last night were just a dream? Hawks is probably legit dead then, sorry bros and the fake chapter was the most predictable shit so you didn&#039;t miss much",
        "name": "Anonymous",
        "no": 200721249,
        "now": "03/25/20(Wed)12:52:09",
        "resto": 200719599,
        "time": 1585155129,
      },
    ],
    "md5": "L3SicznGsKcXH/aXR2Cz4A==",
    "name": "Anonymous",
    "no": 200719599,
    "now": "03/25/20(Wed)12:04:45",
    "omitted_images": 0,
    "omitted_posts": 2,
    "page": 2,
    "replies": 7,
    "resto": 0,
    "semantic_url": "boku-no-hero-academia",
    "sub": "Boku No Hero Academia",
    "tim": 1585152285572,
    "time": 1585152285,
    "tn_h": 249,
    "tn_w": 114,
    "w": 383,
  },
}
`;
